### PR TITLE
Load GO terms from Interpro

### PR DESCRIPTION
**Pull request by bradfordcondon**
- GO analysis displayed only for organisms with results.

- Organism level GO Chart and Trees now display.

- New feature to load GO terms from interpro analysis.  This features allows user to select an interpro analysis instead of uploading a GAF file.  An [interpro to GO mapping file provided by EBI](http://www.ebi.ac.uk/GOA/InterPro2GO) is used to create an index of interpro to GO mappings, and features with those interpro terms have corresponding GO terms inserted into CHADO.

Thank you

-The Staton Lab

### CODE REVIEW

**Comment by spficklin** [comment]
Thanks for submitting!  I have a few questions...

1)  Having the IPR to GO mapping file as a static file means that the IPR -> GO mapping can become somewhat stale in a few months if new IPR / GO term associations are made.  Is there a way to automatically update this file?   
2)  The InterPro Analysis module automatically adds IPR terms to the cvterm tables of Chado when the IPR results are imported, and I believe it imports GO terms if they are in the InterProScan results as well.   What is the use case where IPR terms are provided directly to the GO modules. Is it in the GAF file?
3) I've been thinking we need to get rid of the views on the organism page because some organisms have multiple genomes and multiple analyses and it might be more intuitive to have the GO reports on a separate GO analysis page.  Thoughts?

**Comment by monicacecilia** [comment]
Hi, Ms. Nosey Parker here - I'd like to bring in a few spies from the GO team into this ticket, @kltm and @cmungall, in case they have some insight into the questions you are trying to answer.

**Comment by bradfordcondon** [comment]
Hi Stephen,
1)  A good point.  There's versioning in the header of EBI's BLAST2GO mapping.  @almasaeed2010 has suggested extending tripal_cv to include the IPR2GO download option.  We'll look at this soon.

2)  All of our HWG genomes have Interpro analyses loaded, but no associated GO terms in the cvterms table.  What's more, the GO terms in the Interproscan analysis are outdated and match an older version of GO- these annotations toss warnings when we load them in.  Remapping GO terms based on IPR bypasses this.

In short, the use case is specifically for a) IPR is already loaded, but GO terms were not added at time of loading.  b)  GO terms associated with the IPR are outdated.

3)  I agree this is something to consider.  I dont think any of the species on HWG have multiple genomes.  Is there an example you can point to?
That said, the target GO analysis must be selected from the dropdown box to display the chart/graph.  Wouldn't each genome have its own GO analysis?

**Comment by spficklin** [comment]
Hi @bradfordcondon 

I don't think you want to extend the tripal_cv module as that module will become legacy in Tripal v3.  I really think the best place for the IPR2GO is in the Tripal Analysis InterPro module because you're starting with the IPR terms in an effort to add GO terms.  This code could not function without the IPR terms so I think it's best in that module.   The InterPro module doesn't have any built in GO terms.  It simply uses those that are already in an XML export from InterProScan.    If you are getting messages about GO terms not being up to date it would be caused by an older database used by InterProScan or and older version of GO loaded into Tripal.   In Tripal v2.0 we noticed that our URL for the GO database was older.  So, most likely if you have GO terms loaded you would have been missing more recent updates.  Unfortunately, in the newer updates to GO there is a cycle in the graph which cases the cvtermpath update to crash.  We're working on that fix, but despite that failure at the end you can still get the most recent GO loaded in Tripal v2.1 and v3-beta.  

As per removing the report from the organism page... there are some species that now have multiple genomes and some created by different groups.  For example, there are at least two genome assemblies for Gossypium hirsutum from two different groups.  When there is only a single genome for a species it can be later updated so you'd have multiple versions of the same genome.  So, I think it makes more sense to have the GO/KEGG/IPR reports appear on an analysis page rather than on an organism page.

**Comment by bradfordcondon** [comment]
Hi Stephen,

> I really think the best place for the IPR2GO is in the Tripal Analysis InterPro module because you're starting with the IPR terms in an effort to add GO terms. 

I see where you're coming from, but where would we add the feature?  It is a function that we want performed on old, loaded interpro analyses.  Would adding it as a feature to the "add Interpro analysis" page work, considering we don't want to have to re-load the Interpro analysis?


> If you are getting messages about GO terms not being up to date it would be caused by an older database used by InterProScan or and older version of GO loaded into Tripal.

Our GO terms are out of date because the analyses were performed against outdated databases.  The GO terms loaded by the original IPR analysis point to deprecated terms that have been replaced.  The goal of the IPR2GO feature is to prevent us from going back and rerunning IPR for all of our loaded datasets.  The use case is specifically when you ran the Interpro analysis a while ago and don't want to have to go back to the original raw data.

Note this is a separate issue from our troubles loading the GO terms into Tripal we're discussing elsewhere...

>So, I think it makes more sense to have the GO/KEGG/IPR reports appear on an analysis page rather than on an organism page.

Is this the right example, [Gossypium hirsutum](https://www.cottongen.org/node/503)?  It looks like the two separate assemblies are hosted as separate organisms?  

To clarify, instead of having a dropdown that lets you select a GO analysis, you would want... a table at the organism page with links to each GO analysis, and the chart is displayed on that separate page?   If not that... then how does the user find the GO analysis page?


So to summarize you'd like us to remove the IPR2GO function from this module, and migrate the the GO reports from the organism page to an analysis page before we can merge?  If that's right we'll work on resubmitting.


Thanks

Bradford

**Comment by spficklin** [comment]
Hi @bradfordcondon 

I think you can place your form for importing/updating the GO associations for feature with IPR terms as a new Chado Data Loader.  So the menu path you can use is admin/tripal/loaders/[loader name].  You could call it something like "Update InterPro to GO associations".  In Tripal v3 we have already updated the IPR/Blast/KEGG analysis modules and they no longer import data as part of an analysis.  So, going this route would make it easier for Tripal v3 as well.  

For that cotton genome, there is only one organism page but multiple assemblies.  Each genome assembly has it's own analysis page.  This can happen quite a bit especially if there is an update to an existing assembly (e.g. v1, v2, v3, etc.)  But in this case there were two assemblies of the same "D" genome of cotton, one by BGI and another by JGI and both are hosted on the site.  We will deprecate the display of the GO reports on the organism pages for the Tripal v3 compatible version.  The GO reports will be available on the GO analysis page, and we can make those analyses appear as links on the organism page.   Any design ideas/suggestions on this approach would be welcome.

**To summarize:**

- Yes, move your code for updating GO records associated with IPR terms over to the InterPro Analysis module.
- Automate the download of the IPR2GO.txt file each time your loader runs and remove it from the git repository so that there's no risk of it being stale.
- No need for you to move the GO reports unless you want to get involved in adjusting the Tripal v3 version of the GO analysis module.

**Comment by spficklin** [comment]
Wow, this pull request has been sitting for quite a while.  @bradfordcondon is it still relevant?

**Comment by bradfordcondon** [comment]
Haha honestly stephen i've totally forgotten about it.

* GO analysis displayed only for organisms with results.
* Organism level GO Chart and Trees now display.

These both sound good to me!  I know that I dont have these on the HWG live site right now... Have they since been fixed?  

>I really think the best place for the IPR2GO is in the Tripal Analysis InterPro module because you're starting with the IPR terms in an effort to add GO terms.

Sounds like I needed to move this to the interpro module, remove it from the PR and resubmit.

**Comment by spficklin** [comment]
Okay. I'm going to close it.  But if you change your mind please reopen.
